ENH: Enable short_caption in to_latex

ivanovmg opened the pull request:
- [x] closes #36267
- [x] tests added / passed
- [x] passes black pandas
- [x] passes git diff upstream/master -u -- "*.py" | flake8 --diff
- [x] whatsnew entry


Enable short_caption for ``DataFrame.to_latex`` by expanding the meaning of kwarg ``caption``.
Optionally ``caption`` can be a ``Tuple[str, str] = full_caption, short_caption``.

The final caption macro would look like this:
```
\caption[short_caption]{full_caption}
```

ivanovmg committed:
Extract helper method for caption & label macro

Method _compose_caption_and_label_macro unifies
creation of caption and label macros for both tabular
and longtable envs.

ivanovmg committed:
Enable short_caption for df.to_latex

Kwarg short_caption allows one to add short caption
to LaTeX \caption macro.

The final caption macro would look like this:
```
\caption[short_caption]{caption}
```

ivanovmg committed:
Replace unwanted pytest.warns with tm.assert...

ivanovmg committed:
Fix missing f-string placeholder

ivanovmg committed:
Apply black

simonjayhawkins commented on pandas/core/generic.py:
```diff
@@ -2966,7 +2967,10 @@ def to_latex(
+        short_caption : str, optional
+            The LaTeX short caption.
+            Full caption output would look like this:
+            ``\caption[short_caption]{caption}``.
```
needs a versionadd tag

ivanovmg replied:
Should I put this?
```
.. versionadded:: 1.2.0
```

I was not sure about the version.

simonjayhawkins commented on pandas/core/generic.py:
```diff
@@ -2966,7 +2967,10 @@ def to_latex(
 
             .. versionadded:: 1.0.0
-
+        short_caption : str, optional
```
rather than another keyword parameter, maybe caption could be changed to accept Optional[Union[str, Tuple[str,str]]] to allow the short caption to be passed.

ivanovmg replied:
I can do you this if you think this is the best way.
My concern is that in this case caption kwarg will be too long (another concern is that further unpacking into caption and short_caption from either string or tuple will look imperfect, but it is manageable).

ivanovmg replied:
@simonjayhawkins, can you please confirm that this is the option to proceed with?

simonjayhawkins replied:
i've added the api design and needs discussion labels. we should get some opinions from others before proceeding. Changing the api after the event is not so easy. cc @WillAyd @jreback  @toobaz

toobaz replied:
I wouldn't bother about the caption kwarg being too long, while I do find that the additional argument _might_ be slightly easier to document. But all in all I have a slight preference for avoiding the new argument (and the new error message) and resorting to the tuple version.
I would actually phrase the docs as something like "a tuple (short_caption, full_caption), which will result in \caption[short_caption]{caption}; if a single string is passed, no short caption will be set".

ivanovmg replied:
Should I wait for the decision from the remaining stakeholders?

WillAyd replied:
I don't have a strong preference. Looks like everyone so far has preferred the tuple version though so I'd say go ahead with that

There is some ambiguity as to whether "xy" is a single caption or two if you unpack. Maybe not a big deal but worth calling out

ivanovmg replied:
I made the changes required, including testing corner cases like "xy" caption.

simonjayhawkins reviewed (commented):
Thanks @ivanovmg for the PR. needs a release note (enhancements in 1.2)

ivanovmg committed:
Optionally unpack caption=(caption, short_caption)

There was a discussion after pull request in regards
to the new kwarg ``short_caption``.
It was decided not to introduce the new kwarg into
method ``pd.DataFrame.to_latex()``,
but rather optionally unpack caption into a tuple
(caption, short_caption).

So, if caption = str, then short_caption is None
and caption macros will look like this:
```
\caption{caption}
```

If caption = (long_caption, short_caption),
then caption macros will look like this:
```
\caption[short_caption]{caption}
```

ivanovmg committed:
Merge branch 'master' into feature/latex-shortcaption

ivanovmg committed:
Add edge cases for caption testing

ivanovmg committed:
Pass through black

ivanovmg commented:
For some reason CI / Checks (pull_request) failed with the typing validation.

```
Run source activate pandas-dev
mypy --version
mypy 0.730
Performing static analysis using mypy
pandas/core/generic.py:3032: error: Argument 1 to "DataFrameFormatter" has incompatible type "NDFrame"; expected "DataFrame"  [arg-type]
Found 1 error in 1 file (checked 1030 source files)
Performing static analysis using mypy DONE
##[error]Process completed with exit code 1.
```

But I never touched DataFrameFormatter myself. Can you help?

WillAyd commented on pandas/core/generic.py:
```diff
@@ -2887,7 +2887,8 @@ def to_latex(
         multirow=None,
-        caption=None,
+        caption: Optional[Union[str, Tuple[str, str]]] = None,
+        short_caption=None,
```
This can be removed from the signature right?

ivanovmg replied:
Did that, but the build failed due to exceeding time of 1 hour.

simonjayhawkins replied:
restarted

ivanovmg replied:
Failure of ``TestRegistration.test_pandas_plots_register``, which does not seem to be related to the commits.

ivanovmg committed:
Remove typing and short_caption from to_latex

ivanovmg commented:
Please put this PR on hold before the PR related to full refactoring of ``latex.py`` is merged https://github.com/pandas-dev/pandas/pull/35872 (unless abandoned).
It will be easy to implement the changes of the present PR if https://github.com/pandas-dev/pandas/pull/35872 is merged first.

ivanovmg committed:
Merge branch 'master' into feature/latex-shortcaption

ivanovmg committed:
DOC: add parameters to LatexFormatter docstring

jreback commented on pandas/core/generic.py:
```diff
@@ -3123,6 +3124,18 @@ def to_latex(
         if multirow is None:
             multirow = config.get_option("display.latex.multirow")
 
+        if caption:
```
instead of doing this here, can you do it in the formatter itself (in the constructor is ok)

ivanovmg replied:
Done.

jreback commented on pandas/core/generic.py:
```diff
@@ -3074,11 +3074,12 @@ def to_latex(
             from the pandas config module.
-        caption : str, optional
-            The LaTeX caption to be placed inside ``\caption{}`` in the output.
+        caption : str, tuple, optional
```
str or tuple, optional (change the signature as well)
Optional[Union[str, Tuple[str, str]]]

ivanovmg replied:
Done. I updated typing on caption in DataFrameFormatter.to_latex() and LatexFormatter.

jreback commented on pandas/io/formats/format.py:
```diff
@@ -931,6 +931,7 @@ def to_latex(
         multicolumn_format: Optional[str] = None,
         multirow: bool = False,
         caption: Optional[str] = None,
+        short_caption: Optional[str] = None,
```
same

ivanovmg replied:
Done.

ivanovmg committed:
TYP: remove type ignore for column_format

ivanovmg committed:
REF: move short caption parsing to LatexFormatter

WillAyd commented on pandas/io/formats/latex.py:
```diff
@@ -657,7 +694,27 @@ def _select_builder(self) -> Type[TableBuilderAbstract]:
+                    self._caption, self.short_caption = caption
+                except ValueError as err:
+                    msg = "caption must be either str or tuple of two strings"
+                    raise ValueError(msg) from err
```
Can you add a test that hits this?

ivanovmg replied:
It is there, if you mean raising ValueError.

String 561.
```
        # test that wrong number of params is raised
        with pytest.raises(ValueError):
            df.to_latex(caption=(the_caption, the_short_caption, "extra_string"))
```

ivanovmg replied:
I updated the test to ensure that the error message matches expectations.

ivanovmg commented:
I guess, I need to add release notes into Enhancements section, as @simonjayhawkins highlighted. Would it be v1.2.0?

WillAyd commented:
Correct - 1.2 whatsnew

ivanovmg committed:
DOC: add whatsnew for position and short caption

ivanovmg committed:
DOC: add issue number for short caption

ivanovmg committed:
Merge branch 'master' into feature/latex-shortcaption

ivanovmg committed:
CLN: update error message

ivanovmg committed:
TST: ensure that error message is tested

ivanovmg committed:
TST: add tests for bad tuples

ivanovmg committed:
Merge branch 'master' into feature/latex-shortcaption

simonjayhawkins commented on pandas/core/generic.py:
```diff
@@ -3078,11 +3078,12 @@ def to_latex(
+            which results in \caption[short_caption]{full_caption};
+            if a single string is passed, no short caption will be set.
 
             .. versionadded:: 1.0.0
```
```suggestion
            .. versionchanged:: 1.2.0
```

ivanovmg replied:
But I think I should keep ``.. versionadded`` for the consistency. Shouldn't I?

simonjayhawkins replied:
sure.

can have both. and add a one-liner describing the changes.

ivanovmg replied:
Added these tags for caption, position (PR https://github.com/pandas-dev/pandas/pull/35284) and ``to_latex`` function itself.

simonjayhawkins commented on pandas/io/formats/latex.py:
```diff
@@ -605,18 +641,18 @@ def __init__(
         self.multicolumn_format = multicolumn_format
         self.multirow = multirow
-        self.caption = caption
+        self.caption = caption  # type: ignore[assignment]
```
can you add the mypy message that this silences as a comment

ivanovmg replied:
@simonjayhawkins, can you please explain a little bit? I already added ``# type: ignore [assignment]`` to silence mypy error. Is there any other type of silencing? Mypy seem to pass correctly.

simonjayhawkins replied:
to help readers of the code know what the problem is. (majority of #type: ignores in codebase should have comments)

ivanovmg replied:
Oh, I see, ok.

ivanovmg replied:
Done

WillAyd replied:
I am a little confused by this though - are you sure this required? `str` should be fine on the right hand side of an assignment based off of the comment.

ivanovmg replied:
Mypy throws error if it is not there. I will get rid of caption setter (as @jreback suggested), so there will be no issue with mypy error. For some reason I noticed that mypy throws errors when setter accepts ``Optional[something]`` and assigns ``something`` to an attribute (without optional).

ivanovmg committed:
DOC: add/update versionadded, versionchanged tags

ivanovmg committed:
TST: add assertions in caption setter to help mypy

ivanovmg committed:
DOC: add reason for caption type ignore

ivanovmg committed:
DOC: add reason for strrows arg-type ignore

ivanovmg committed:
DOC: add missing empty line before versionadded

ivanovmg committed:
Merge branch 'master' into feature/latex-shortcaption

ivanovmg commented:
I resolved conflicts with the master branch. It seems that there is an issue with the new style docstring (https://github.com/pandas-dev/pandas/pull/33633). ``Returns`` section gets over indented (as a result of running ``python scripts/validate_docstrings.py pandas.DataFrame.to_latex``).

ivanovmg commented:
@jreback, @WillAyd, @simonjayhawkins, @toobaz, ping. Is it good to go?

WillAyd commented on pandas/tests/io/formats/test_to_latex.py:
```diff
@@ -584,6 +653,31 @@ def test_to_latex_longtable_caption_label(self):
+"""
+        assert result_cl == expected_cl
+
+        # test when the caption, the short_caption and the label are provided
```
Same comment here - modular tests are much easier to debug in case of future issues

ivanovmg replied:
I absolutely agree. I would prefer this https://github.com/pandas-dev/pandas/pull/36528 to be merged first, and then rebase on top of that with the separate tests.

ivanovmg replied:
Splitted tests.

WillAyd commented on pandas/tests/io/formats/test_to_latex.py:
```diff
@@ -515,12 +516,80 @@ def test_to_latex_caption_label(self):
+"""
+        assert result_cl == expected_cl
+
+        # test when the short_caption is provided alongside caption
```
Can you split each comment here into a separate test?

ivanovmg replied:
Did that.

jreback commented on pandas/io/formats/latex.py:
```diff
@@ -41,6 +41,12 @@ def __init__(
         self.clinebuf: List[List[int]] = []
         self.strcols = self._get_strcols()
+
+        # Here is a reason for ignoring typing.
```
what is the output of _get_strcols()?

ivanovmg replied:
``_get_strcols()`` returns List[List[str]], where the inner list contains the elements of the column.

jreback commented on pandas/io/formats/latex.py:
```diff
@@ -657,7 +706,29 @@ def _select_builder(self) -> Type[TableBuilderAbstract]:
 
     @property
-    def column_format(self) -> str:
+    def caption(self) -> str:
```
is there a reason we use setters / getters rather than just typing in teh class? seems much simpler.

ivanovmg replied:
Seems that there is no particular reason as the class is supposed to be initialized only once. If this is so, then I will update this into a function, which is run at init.

ivanovmg committed:
REF: replace caption setter with initialize method

ivanovmg committed:
Merge branch 'master' into feature/latex-shortcaption

ivanovmg committed:
TST: align longtable test with the recent changes

Test for longtable with shorcaption was changed
to reflect the recent changes on master branch.
The changes in longtable environment were introduced recently,
see GH #34360.

ivanovmg committed:
Merge branch 'master' into feature/latex-shortcaption

ivanovmg committed:
TST: add for list [full_caption, short_caption]

WillAyd commented on pandas/io/formats/latex.py:
```diff
@@ -384,8 +388,23 @@ def _position_macro(self) -> str:
+            \caption{caption_string}.
+        """
+        if self.caption:
+            return f"\\caption{self._short_caption_macro}{{{self.caption}}}"
```
```suggestion
            return f"\\caption{self._short_caption or ''}{{{self.caption}}}"
```

This can be simplified to get rid of the property

ivanovmg replied:
``short_caption_macro`` is either an empty string or ``[short caption text]`` (in square brackets).
Putting this logic into f-string seems to be rather complicated.
IMHO having a separate property (``_short_caption_macro``) with clear intent is more readable.

ivanovmg replied:
@WillAyd, does my answer make sense?
@jreback, @simonjayhawkins, @toobaz, any plans on merging this?
Or does it require additional work?

jreback replied:
@ivanovmg am looking now, we have 200 open PRs; these take time

jreback replied:
the way you have it is ok; the f-string actually is pretty complicated to grok on the substitutions. if you can make it simpler would take (e.g. use concatenation maybe)

ivanovmg replied:
I made string joining instead of long f-string. Would you consider it more readable?

WillAyd commented on pandas/io/formats/latex.py:
```diff
@@ -384,8 +388,23 @@ def _position_macro(self) -> str:
+            return f"\\caption{self._short_caption_macro}{{{self.caption}}}"
+        return ""
+
+    @property
```
See above - I think can remove this altogether

ivanovmg committed:
Merge branch 'master' into feature/latex-shortcaption

ivanovmg committed:
Merge branch 'master' into feature/latex-shortcaption

jreback commented on pandas/io/formats/latex.py:
```diff
@@ -670,8 +707,26 @@ def _select_builder(self) -> Type[TableBuilderAbstract]:
+                    msg = "caption must be either a string or a tuple of two strings"
+                    raise ValueError(msg) from err
+        else:
+            long_caption = ""
```
can you make this a free function

ivanovmg replied:
I moved this function to module level.

jreback reviewed (changes requested):
small comments, ping on green

ivanovmg committed:
Merge branch 'master' into feature/latex-shortcaption

ivanovmg committed:
REF: use string concat for caption macro

ivanovmg committed:
REF: move method to module level

ivanovmg committed:
REF: drop property _short_caption_macro

ivanovmg commented:
@jreback all checks are good except Windows py38_np18, which failed to create anaconda environment.

jreback commented on pandas/io/formats/latex.py:
```diff
@@ -596,15 +648,32 @@ def env_end(self) -> str:
+        if a single string is passed, no short caption will be set.
+    label : str, optional
+        The LaTeX label to be placed inside ``\label{}`` in the output.
+    position : str, optional
```
is this validated? what are the positbilities here? can you enumerate

jreback replied:
or is this a latex positioning string and we don't need / want to validate explicity?

ivanovmg replied:
Yes, it is LaTeX float positioning. I think it may be a good idea to add validation in a follow-up PR. Not only for position, but also for ``column_format`` (which is the alignment of the columns).

ivanovmg committed:
Merge branch 'master' into feature/latex-shortcaption

ivanovmg commented:
All checks are good except Windows:

```
##[error]We stopped hearing from agent Azure Pipelines 8. 
```

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -96,6 +96,32 @@ For example:
+   print(table)
+
+Usage of keyword ``caption`` is extended.
+Besides taking a single string as an argument,
```
I think we could add a link to to_latest here: https://pandas.pydata.org/docs/user_guide/io.html (followon as prob need a short section as well).

jreback reviewed (approved):
lgtm ping on green.

jreback commented:
thanks @ivanovmg
